src/ucp/core/ucp_listener.c

```diff
@@ -195,76 +195,111 @@ ucs_status_t ucp_listener_query(ucp_listener_h listener, ucp_listener_attr_t *at
-        ucp_worker_iface_cleanup(&listener->wifaces[i]);
+        ucp_worker_iface_cleanup(&listener->ifaces.wifaces[i]);
     }
 
-    ucs_free(listener->wifaces);
+    ucs_free(listener->ifaces.wifaces);
 }
 
-ucs_status_t ucp_listener_create(ucp_worker_h worker,
-                                 const ucp_listener_params_t *params,
-                                 ucp_listener_h *listener_p)
+static ucs_status_t
+ucp_listen_on_cm(ucp_listener_h listener, const ucp_listener_params_t *params)
 {
-    ucp_context_h context   = worker->context;
-    ucp_listener_h listener = NULL;
-    int sockaddr_tls        = 0;
-    char saddr_str[UCS_SOCKADDR_STRING_LEN];
-    ucp_tl_resource_desc_t *resource;
-    uct_iface_params_t iface_params;
-    ucp_worker_iface_t *tmp;
-    ucp_rsc_index_t tl_id;
+    ucp_worker_h worker           = listener->worker;
+    const ucp_rsc_index_t num_cms = ucp_worker_get_cm_num(worker);
+    uct_listener_params_t uct_params;
+    ucp_rsc_index_t i, count;
+    char addr_str[UCS_SOCKADDR_STRING_LEN];
     ucs_status_t status;
-    ucp_tl_md_t *tl_md;
-    uint16_t port;
-    int i;
 
-    if (!(params->field_mask & UCP_LISTENER_PARAM_FIELD_SOCK_ADDR)) {
-        ucs_error("Missing sockaddr for listener");
-        return UCS_ERR_INVALID_PARAM;
+    uct_params.field_mask      = UCT_LISTENER_PARAM_FIELD_CONN_REQUEST_CB |
+                                 UCT_LISTENER_PARAM_FIELD_USER_DATA;
+    uct_params.conn_request_cb = (void *)0xdeadbeaf; /* TODO: ucp_listener_conn_request_cb; */
+    uct_params.user_data       = listener;
+
+    listener->cms.listeners = ucs_calloc(num_cms,
+                                         sizeof(*listener->cms.listeners),
+                                         "uct_listeners_arr");
+    if (listener->cms.listeners == NULL) {
+        ucs_error("Can't allocate memory for UCT listeners array");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    UCP_CHECK_PARAM_NON_NULL(params->sockaddr.addr, status, return status);
-
-    if (ucs_test_all_flags(params->field_mask,
-                           UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER |
-                           UCP_LISTENER_PARAM_FIELD_CONN_HANDLER)) {
-        ucs_error("Only one accept handler should be provided");
-        return UCS_ERR_INVALID_PARAM;
+    for (i = 0, count = 0; i < num_cms; ++i) {
+        status = uct_listener_create(worker->cms[i], params->sockaddr.addr,
+                                     params->sockaddr.addrlen, &uct_params,
+                                     &listener->cms.listeners[count]);
+        if (status == UCS_OK) {
+            ++count;
+        } else {
+            ucs_debug("uct_listener_create failed on CM %p with address %s",
+                      worker->cms[i],
+                      ucs_sockaddr_str(params->sockaddr.addr, addr_str,
+                                       ucs_static_array_size(addr_str)));
+            ucs_assert(status == UCS_ERR_INVALID_ADDR);
```

Comment:
why only this status is valid?
let's have a graceful handling in case of other errors - not an assert...